File: server/src/main/java/rdp/RDPDatagram.java
package rdp;

import ttp.TTPPacket;
import util.Utility;

public class RDPDatagram {
	
	// List of headers and data field
	private int sequence;
	private int acknowledgement;
	private int head;
	private int tail;
	private int length;
	private byte[] data;
	
	// Datagram sequence number tracker
	private static int currentSequenceNumber = 1;
	
	public RDPDatagram(byte[] bytes) {
		this.sequence = Utility.byteToInt(bytes, 0);
		this.acknowledgement = Utility.byteToInt(bytes, 4);
		this.head = Utility.byteToInt(bytes, 8);
		this.tail = Utility.byteToInt(bytes, 12);
		this.length = Utility.byteToInt(bytes, 16);
		this.data = Utility.splitBytes(bytes, 20, 20 + this.length);
	}
	
	public RDPDatagram(int acknowledgement, int head, int tail, TTPPacket packet) {
		this.sequence = getNextSequenceNumber();
		this.acknowledgement = acknowledgement;
		this.head = head;
		this.tail = tail;
		if (packet == null) this.data = new byte[0];
		else this.data = packet.getBytes();
		this.length = this.data.length;
	}
	
	public RDPDatagram(int acknowledgement, int fragDistance, TTPPacket packet) {
		this.sequence = getNextSequenceNumber();
		this.acknowledgement = acknowledgement;
		this.head = this.sequence;
		this.tail = this.head + fragDistance;
		this.data = packet.getBytes();
		this.length = this.data.length;
	}
	
	public int getSequence() {
		return sequence;
	}

	public int getAcknowledgement() {
		return acknowledgement;
	}

	public int getHead() {
		return head;
	}

	public int getTail() {
		return tail;
	}
	
	public int getLength() {
		return this.length;
	}

	public byte[] getData() {
		return data;
	}
	
	public byte[] getBytes() {
		return Utility.mergeBytes(new byte[][] {
			Utility.intToByte(this.sequence),
			Utility.intToByte(this.acknowledgement),
			Utility.intToByte(this.head),
			Utility.intToByte(this.tail),
			Utility.intToByte(this.length),
			this.data});
	}
	
	public TTPPacket getTTPPacket() {
		return new TTPPacket(this.data);
	}
	
	private static int getNextSequenceNumber() {
		if (currentSequenceNumber == Integer.MAX_VALUE) {
			currentSequenceNumber = 1;
			return Integer.MAX_VALUE;
		}
		return currentSequenceNumber++;
	}
	
	public static void resetSequenceIfLarger(int size) {
		if (currentSequenceNumber + size > Integer.MAX_VALUE) currentSequenceNumber = 1;
	}

	@Override
	public String toString() {
		StringBuilder builder = new StringBuilder();
		builder.append("RDPDatagram [sequence=");
		builder.append(sequence);
		builder.append(", acknowledgement=");
		builder.append(acknowledgement);
		builder.append(", head=");
		builder.append(head);
		builder.append(", tail=");
		builder.append(tail);
		builder.append(", length=");
		builder.append(length);
		if (length > 0) {
			builder.append(", data=");
			builder.append(getTTPPacket().toString());
		}
		return builder.toString();
	}

}

File: python_client/RDP_datagram.py
import utility
import TTP_packet as ttp

MAX_INTEGER = 2147483647

current_sequence_number = 1


def get_next_sequence_number():
    global current_sequence_number
    global MAX_INTEGER
    if current_sequence_number == MAX_INTEGER:
        current_sequence_number = 1
        return MAX_INTEGER
    else:
        current_sequence_number += 1
        return current_sequence_number - 1


def reset_sequence_if_larger(size):
    global current_sequence_number
    global MAX_INTEGER
    if current_sequence_number + size > MAX_INTEGER:
        current_sequence_number = 1


class RDP_Datagram():
    def __init__(self, bytes_arr=[], acknowledgement=0, head=0, tail=0, frag_distance=None, packet=None):
        if len(bytes_arr) > 0:
            self.sequence = utility.byte_to_int(bytes_arr, 0)
            self.acknowledgement = utility.byte_to_int(bytes_arr, 4)
            self.head = utility.byte_to_int(bytes_arr, 8)
            self.tail = utility.byte_to_int(bytes_arr, 12)
            self.length = utility.byte_to_int(bytes_arr, 16)
            self.data = bytes_arr[20:20 + self.length]
        else:
            self.sequence = get_next_sequence_number()
            self.acknowledgement = acknowledgement
            if packet is None:
                self.data = b''
            else:
                self.data = packet.get_bytes()
            if frag_distance is None:
                self.head = head
                self.tail = tail
            else:
                self.head = self.sequence
                self.tail = self.head + frag_distance
            self.length = len(self.data)

    def get_bytes(self):
        return b"".join([
                        utility.int_to_byte(self.sequence),
                        utility.int_to_byte(self.acknowledgement),
                        utility.int_to_byte(self.head),
                        utility.int_to_byte(self.tail),
                        utility.int_to_byte(self.length),
                        self.data])

    def get_TTP_packet(self):
        return ttp.TTP_packet(bytes_arr=self.data)